cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,262 @@
+  {
+    return std::sqrt(in);
+  }
+};
+
+struct nz_op {
+  template <typename Type, typename IdxType = int>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type in, IdxType i = 0) const
+  {
+    return in != Type(0) ? Type(1) : Type(0);
+  }
+};
+
+struct abs_op {
+  template <typename Type, typename IdxType = int>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type in, IdxType i = 0) const
+  {
+    return std::abs(in);
+  }
+};
+
+struct sq_op {
+  template <typename Type, typename IdxType = int>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type in, IdxType i = 0) const
+  {
+    return in * in;
+  }
+};
+
+struct add_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a + b;
+  }
+};
+
+struct sub_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a - b;
+  }
+};
+
+struct mul_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a * b;
+  }
+};
+
+struct div_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a / b;
+  }
+};
+
+struct div_checkzero_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b == Type{0}) { return Type{0}; }
+    return a / b;
+  }
+};
+
+struct pow_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    return std::pow(a, b);
+  }
+};
+
+struct min_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (a > b) { return b; }
+    return a;
+  }
+};
+
+struct max_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(ScalarT s) : scalar{s} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(InT unused) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ComposedOpT, typename ScalarT>
```

Comment:
Would you consider swapping the template arguments here? I have a feeling there's a greater chance one would want to force the type of the scalar, but keep the `ComposedOpT` inferred (then, perhaps the order of the field should be swapped as well). Also, maybe rename the latter to `BinaryOpT`, so that it's clear we expect the binary operator as the argument.

Reply:
I don't see how `ComposedOpT` could be inferred. The only argument to the constructor is the scalar, so if C++ was a bit smarter it could infer `ScalarT` using C++17's CTAD, but unfortunately, the other template argument seems to break that.

Reply:
I'll see about having two constructors, then. The one-arg constructor is convenient for the aliases:

```c++
template <typename Type>
using scalar_add_op = scalar_op<add_op, Type>;

template <typename Type>
using scalar_sub_op = scalar_op<sub_op, Type>;

...
```

However, I'm still unsure in which case `ComposedOpT` will be inferred and `ScalarT` won't? With a two-arg constructor, both are inferred.

Reply:
I'm assuming the case when the scalar you provide is of different type and you want it be implicitly converted.

Reply:
Nothing that can't be achieved with a cast:

```c++
uint32_t offset;
scalar_op<size_t> op1(add_op, offset);
scalar_op op2(add_op, (size_t)offset);
```

I'm wondering if, in a future version, C++ will be clever enough to infer `ScalarT` when constructing like:

```c++
scalar_op<add_t> op(offset);
```

Reply:
Sure it can be achieved with the cast; it's just my personal preference to like typing `scalar_op<size_t>(1, f)` instead of  `scalar_op(f, size_t(1))`.
Btw, c-style casts are discouraged by our clang-format; hence in your case one would either use constructor `scalar_op op2(add_op, size_t(offset));` or static cast `scalar_op op2(add_op, static_cast<size_t>(offset))`.